# **Question 1**

**MNIST Dataset**

In [ ]:
import tensorflow as tf
import numpy as np

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Take one random image from each class
images = []
labels = []
for i in range(10):
  idx = np.random.choice(np.where(y_train == i)[0])
  images.append(x_train[idx])
  labels.append(i)

# Estimate L1 distance to every other class
distances = np.zeros((10, 10))
for i in range(10):
  for j in range(10):
    distances[i, j] = np.mean(np.abs(images[i] - images[j]))

# Display the results with corresponding classes
print("L1 distances between random images from each class:")
print("--------------------------------------------------")
for i in range(10):
  print(f"From Random {labels[i]}:")
  for j in range(10):
    print(f"  To class {j}: {distances[i, j]}")
  print()

# Find the closest neighboring class (K=1) based on L1 distance for each image class
closest_classes = np.argmin(distances, axis=1)

# Display the closest neighboring classes
print("Closest neighboring class (K=1) based on L1 distance for each image class:")
print("---------------------------------------------------------------------------")
for i in range(10):
    print(f"Random image  {labels[i]} is closest to Class {closest_classes[i]}")

# Estimate L2 distance to every other class
distances_l2 = np.zeros((10, 10))
for i in range(10):
    for j in range(10):
        distances_l2[i, j] = np.sqrt(np.mean(np.square(images[i] - images[j])))


# Display the results with corresponding classes
print("L2 distances between random images from each class:")
print("--------------------------------------------------")
for i in range(10):
    print(f"From Rnadom {labels[i]}:")
    for j in range(10):
        print(f"  To class {j}: {distances_l2[i, j]}")
    print()

# Find the closest neighboring class (K=1) based on L2 distance for each image class
closest_classes_l2 = np.argmin(distances_l2, axis=1)

# Display the closest neighboring classes
print("Closest neighboring class (K=1) based on L2 distance for each image class:")
print("---------------------------------------------------------------------------")
for i in range(10):
    print(f"Random image {labels[i]} is closest to Class {closest_classes_l2[i]}")

L1 distances between random images from each class:
--------------------------------------------------
From Random 0:
  To class 0: 0.0
  To class 1: 33.50765306122449
  To class 2: 32.8125
  To class 3: 38.53826530612245
  To class 4: 34.30994897959184
  To class 5: 37.692602040816325
  To class 6: 33.56122448979592
  To class 7: 39.891581632653065
  To class 8: 37.48979591836735
  To class 9: 39.21683673469388

From Random 1:
  To class 0: 39.96173469387755
  To class 1: 0.0
  To class 2: 32.610969387755105
  To class 3: 30.1734693877551
  To class 4: 27.904336734693878
  To class 5: 29.32780612244898
  To class 6: 30.09438775510204
  To class 7: 32.83290816326531
  To class 8: 26.839285714285715
  To class 9: 32.81122448979592

From Random 2:
  To class 0: 39.02423469387755
  To class 1: 39.22576530612245
  To class 2: 0.0
  To class 3: 41.97066326530612
  To class 4: 34.150510204081634
  To class 5: 39.81887755102041
  To class 6: 40.25892857142857
  To class 7: 42.99744897959184
 

*The MNIST dataset contains handwritten digit images, where shared features such as strokes and loops may lead to similarities across different digits. Certain digits, like 3 and 8, may inherently share visual characteristics, potentially causing confusion during classification. Limited variation within each digit class can result in images appearing similar due to constrained diversity in writing styles. Misclassification may arise when algorithms focus on less discriminative features, deviating from human perception of digit similarity.*

**CIFAR-10 Dataset**

In [ ]:
import tensorflow as tf
import numpy as np

# Load the CIFAR-10 dataset
cifar10 = tf.keras.datasets.cifar10
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

# Take one random image from each class
images = []
labels = []
for i in range(10):
    idx = np.random.choice(np.where(y_train[:,0] == i)[0])
    images.append(x_train[idx])
    labels.append(i)

# Estimate L1 distance to every other class
distances = np.zeros((10, 10))
for i in range(10):
    for j in range(10):
        distances[i, j] = np.mean(np.abs(images[i] - images[j]))

# Display the results with corresponding classes
print("L1 distances between random images from each class:")
print("--------------------------------------------------")
for i in range(10):
    print(f"From Random {labels[i]}:")
    for j in range(10):
        print(f"  To class {j}: {distances[i, j]}")
    print()
# Find the closest neighboring class (K=1) based on L1 distance for each image class
closest_classes = np.argmin(distances, axis=1)

# Display the closest neighboring classes
print("Closest neighboring class (K=1) based on L1 distance for each image class:")
print("---------------------------------------------------------------------------")
for i in range(10):
    print(f"Random image {labels[i]} is closest to Class {closest_classes[i]}")

# Estimate L2 distance to every other class
distances_l2 = np.zeros((10, 10))
for i in range(10):
    for j in range(10):
        distances_l2[i, j] = np.sqrt(np.mean(np.square(images[i] - images[j])))

# Display the results with corresponding classes
print("L2 distances between random images from each class:")
print("--------------------------------------------------")
for i in range(10):
    print(f"From random {labels[i]}:")
    for j in range(10):
        print(f"  To class {j}: {distances_l2[i, j]}")
    print()

# Find the closest neighboring class (K=1) based on L2 distance for each image class
closest_classes_l2 = np.argmin(distances_l2, axis=1)

# Display the closest neighboring classes
print("Closest neighboring class (K=1) based on L2 distance for each image class:")
print("---------------------------------------------------------------------------")
for i in range(10):
    print(f"Random image {labels[i]} is closest to Class {closest_classes_l2[i]}")

L1 distances between random images from each class:
--------------------------------------------------
From Random 0:
  To class 0: 0.0
  To class 1: 147.951171875
  To class 2: 123.96907552083333
  To class 3: 127.06901041666667
  To class 4: 120.4580078125
  To class 5: 136.13411458333334
  To class 6: 102.49739583333333
  To class 7: 137.50716145833334
  To class 8: 111.26627604166667
  To class 9: 128.93880208333334

From Random 1:
  To class 0: 107.548828125
  To class 1: 0.0
  To class 2: 145.85123697916666
  To class 3: 149.36783854166666
  To class 4: 154.7568359375
  To class 5: 142.01627604166666
  To class 6: 150.462890625
  To class 7: 137.97265625
  To class 8: 128.06510416666666
  To class 9: 129.57096354166666

From Random 2:
  To class 0: 131.69759114583334
  To class 1: 108.73209635416667
  To class 2: 0.0
  To class 3: 117.7666015625
  To class 4: 124.40559895833333
  To class 5: 132.99837239583334
  To class 6: 155.36165364583334
  To class 7: 111.20475260416667
  To


*The CIFAR-10 dataset comprises images of 10 different classes, where common features like colors and shapes may lead to visual similarities across classes. Certain classes, like birds and airplanes, may share characteristics, contributing to potential confusion. Limited variation within each class can cause images from different classes to appear similar due to constrained diversity. Image misclassification can occur when algorithms prioritize features less discernible to humans, leading to discrepancies in perceived similarity.*

# **Question 2**



In [ ]:
### Import necessary libraries
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
import time
import tensorflow.keras.datasets.mnist

# Load the MNIST dataset
mnist = tensorflow.keras.datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalize pixel values
X_train = X_train / 255.0
X_test = X_test / 255.0

# 1. Divide the MNIST data into train (80%) and test (20%)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# 2. Based on L1 distance function, run K-nearest neighbor classifier for K=5 on test set
start_time = time.time()
knn_l1_k5 = KNeighborsClassifier(n_neighbors=5, metric='manhattan')
knn_l1_k5.fit(X_train.reshape(-1, 28 * 28), y_train)
y_pred_l1_k5 = knn_l1_k5.predict(X_val.reshape(-1, 28 * 28))
end_time = time.time()

# Calculate accuracy and computing time
accuracy_l1_k5 = accuracy_score(y_val, y_pred_l1_k5)
computing_time_l1_k5 = end_time - start_time

print(f"L1 distance, K=5: Accuracy: {accuracy_l1_k5:.3f}, Computing Time: {computing_time_l1_k5:.3f} seconds")

# 3. Repeat steps 2 and 3 for K=25, 55, and 105
k_values = [25, 55, 105]
for k in k_values:
    start_time = time.time()
    knn_l1 = KNeighborsClassifier(n_neighbors=k, metric='manhattan')
    knn_l1.fit(X_train.reshape(-1, 28 * 28), y_train)
    y_pred_l1 = knn_l1.predict(X_val.reshape(-1, 28 * 28))
    end_time = time.time()

    accuracy_l1 = accuracy_score(y_val, y_pred_l1)
    computing_time_l1 = end_time - start_time

    print(f"L1 distance, K={k}: Accuracy: {accuracy_l1:.3f}, Computing Time: {computing_time_l1:.3f} seconds")

# 4. Repeat steps 2,3, and 4 with L2 as a distance function
for k in k_values:
    start_time = time.time()
    knn_l2 = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
    knn_l2.fit(X_train.reshape(-1, 28 * 28), y_train)
    y_pred_l2 = knn_l2.predict(X_val.reshape(-1, 28 * 28))
    end_time = time.time()

    accuracy_l2 = accuracy_score(y_val, y_pred_l2)
    computing_time_l2 = end_time - start_time

    print(f"L2 distance, K={k}: Accuracy: {accuracy_l2:.3f}, Computing Time: {computing_time_l2:.3f} seconds")


L1 distance, K=5: Accuracy: 0.965, Computing Time: 599.459 seconds
L1 distance, K=25: Accuracy: 0.948, Computing Time: 564.934 seconds
L1 distance, K=55: Accuracy: 0.935, Computing Time: 576.635 seconds
L1 distance, K=105: Accuracy: 0.922, Computing Time: 564.188 seconds
L2 distance, K=25: Accuracy: 0.957, Computing Time: 34.299 seconds
L2 distance, K=55: Accuracy: 0.946, Computing Time: 34.742 seconds
L2 distance, K=105: Accuracy: 0.933, Computing Time: 34.289 seconds



*Summary of the key observations:*

*Accuracy:*
*Accuracy tends to decrease as we increase the number of neighbors (K).*
*L2 distance generally leads to slightly higher accuracy compared to L1 distance.*

*Computing Time:*
*L1 distance takes much longer to compute compared to L2 distance.*
*L2 distance is faster and more stable across different K values.*

*Effect of K:Increasing K can lead to higher computational cost without necessarily improving accuracy.*

*Distance Function Comparison:*
*L2 distance consistently performs better in terms of both accuracy and computing time compared to L1 distance.*
*Overall, for this MNIST classification task, L2 distance proves to be more efficient and yields better results than L1 distance.*

# **Question 3**

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.datasets import mnist

# Load and preprocess the MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
train_images, test_images = train_images / 255.0, test_images / 255.0

# Step 1: Implementing the baseline model
model_baseline = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_baseline.compile(optimizer='adam',
                       loss='sparse_categorical_crossentropy',
                       metrics=['accuracy'])

model_baseline.fit(train_images, train_labels, epochs=5)
test_loss, test_acc = model_baseline.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy for baseline model:', test_acc)

# Step 2: Changing the number of neurons in the second layer to the same size as the input
model_same_size = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(28*28, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_same_size.compile(optimizer='adam',
                        loss='sparse_categorical_crossentropy',
                        metrics=['accuracy'])

model_same_size.fit(train_images, train_labels, epochs=5)
test_loss, test_acc = model_same_size.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy for model with same size second layer:', test_acc)

# Step 3: Changing the number of neurons in the second layer to 2 times the size of the input
model_double_size = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(2*28*28, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_double_size.compile(optimizer='adam',
                          loss='sparse_categorical_crossentropy',
                          metrics=['accuracy'])

model_double_size.fit(train_images, train_labels, epochs=5)
test_loss, test_acc = model_double_size.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy for model with double size second layer:', test_acc)

# Step 4: Adding another dense layer with random neurons
model_random_layer = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(128, activation='relu'),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_random_layer.compile(optimizer='adam',
                           loss='sparse_categorical_crossentropy',
                           metrics=['accuracy'])

model_random_layer.fit(train_images, train_labels, epochs=5)
test_loss, test_acc = model_random_layer.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy for model with additional random dense layer:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.2562 - accuracy: 0.9257
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1127 - accuracy: 0.9669
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0780 - accuracy: 0.9765
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0597 - accuracy: 0.9820
Epoch 5/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0457 - accuracy: 0.9857
313/313 - 1s - loss: 0.0805 - accuracy: 0.9749 - 602ms/epoch - 2ms/step

Test accuracy for baseline model: 0.9749000072479248
Epoch 1/5
1875/1875 [==============================] - 17s 9ms/step - loss: 0.1900 - accuracy: 0.9434
Epoch 2/5
1875/1875 [==============================] - 18s 9ms/step - loss: 0.0765 - accuracy: 0.9765
Epoch 3/5
1875/1875 [==============================] - 16s 9ms/step - loss: 0.0486 - accuracy: 0.9848
Epoch 4/5
1875/1875 [==============================] - 16s 8ms/

*The provided training results demonstrate experiments with variations of a baseline neural network model:*

*Baseline Model: Achieved a test accuracy of around 97.49% after 5 epochs*.

*Model Variations:*

*Same Size Second Layer: Reached approximately 97.79% accuracy.*
*Double Size Second Layer: Improved slightly to about 97.92% accuracy.*
*Additional Random Dense Layer: Maintained accuracy at around 97.93%.*

*Overall, these adjustments led to marginal improvements in accuracy without significant changes in training time*

# **Bonus Question**

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from tensorflow.keras.datasets import fashion_mnist
import tensorflow as tf
from tensorflow.keras import layers, models

# Load Fashion MNIST dataset
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Normalize pixel values to [0, 1]
train_images = train_images.astype('float32') / 255
test_images = test_images.astype('float32') / 255

# Split the data into training and validation sets
train_images, val_images, train_labels, val_labels = train_test_split(train_images, train_labels, test_size=0.2, random_state=42)

# Define and train the KNN classifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(train_images.reshape(train_images.shape[0], -1), train_labels)

# Evaluate the KNN classifier on the validation set
accuracy_knn = knn.score(val_images.reshape(val_images.shape[0], -1), val_labels)
print("Accuracy of KNN on validation set:", accuracy_knn)


# Define the MLP model
model = models.Sequential([
    layers.Flatten(input_shape=(28, 28)),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the MLP model
history = model.fit(train_images, train_labels, epochs=10, validation_data=(val_images, val_labels))

# Evaluate the MLP model on the test set
test_loss, test_accuracy = model.evaluate(test_images, test_labels)
print("Accuracy of MLP on test set:", test_accuracy)


Accuracy of KNN on validation set: 0.85275
Epoch 1/10
1500/1500 [==============================] - 6s 3ms/step - loss: 0.5223 - accuracy: 0.8174 - val_loss: 0.4072 - val_accuracy: 0.8568
Epoch 2/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.3894 - accuracy: 0.8599 - val_loss: 0.3831 - val_accuracy: 0.8592
Epoch 3/10
1500/1500 [==============================] - 5s 3ms/step - loss: 0.3470 - accuracy: 0.8734 - val_loss: 0.3795 - val_accuracy: 0.8658
Epoch 4/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.3188 - accuracy: 0.8834 - val_loss: 0.3553 - val_accuracy: 0.8711
Epoch 5/10
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2995 - accuracy: 0.8897 - val_loss: 0.3369 - val_accuracy: 0.8818
Epoch 6/10
1500/1500 [==============================] - 6s 4ms/step - loss: 0.2860 - accuracy: 0.8938 - val_loss: 0.3246 - val_accuracy: 0.8852
Epoch 7/10
1500/1500 [==============================] - 5s 3ms/step - loss: 0.2709 - accuracy


*The KNN model achieved a validation accuracy of approximately 85.3%, while the MLP model achieved a test accuracy of around 88.1%. This indicates that the MLP model outperformed the KNN model on the Fashion MNIST dataset, demonstrating the effectiveness of neural network-based approaches for image classification tasks compared to traditional machine learning methods like KNN.*